src/webhook_launcher/app/payload.py

```diff
@@ -429,3 +442,113 @@ def handle(self):
+            commits = change.get('commits', [])
+            if not newref:
+                # TODO: Handle delete events
+                continue
+            name = newref['name']
+            hash = newref['target']['hash']
+            if newref['type'] == 'branch':
+                branches[name] = [hash, commits]
+            elif newref['type'] == 'tag':
+                tags[name] = [hash, set()]
+            else:
+                print("Unknown ref type '%s'" % newref['type'])
+                continue
+
+        # Find branches for tags
+        for tag, (hash, tag_branches) in tags.items():
+            # TODO: this will only find branches with head matching the tag
+            bbcall = bbAPIcall(self.data['repository']['full_name'])
+            bts = bbcall.branches_tags()
+            for branch in bts['branches']:
+                if branch['changeset'] == hash:
+                    tags[tag][1].add(branch['name'])
+
+        # Handle commits in branches
+        for branch, (revision, commits) in branches.iteritems():
+            mapobjs = WebHookMapping.objects.filter(
+                repourl=self.url, branch=branch,
+            )
+            if not len(mapobjs):
+                packages = self.params.get("packages", None)
+                mapobjs = self.create_placeholder(
+                    self.url, branch, packages=packages
+                )
+            notified = False
+            for mapobj in mapobjs:
+                seenrev, created = LastSeenRevision.objects.get_or_create(
+                    mapping=mapobj
+                )
+
+                emails = set()
+                for commit in commits:
+                    emails.add(commit['author']['raw'])
+                    if len(emails) == 2:
+                        break
+                if emails:
+                    seenrev.emails = json.dumps(list(emails))
+
+                if created or seenrev.revision != revision:
+                    print(
+                        "%s in %s was not seen before, notify it if enabled" %
+                        (revision, branch)
+                    )
+                    seenrev.revision = revision
+                    seenrev.payload = json.dumps(self.data)
+                    handle_commit(
+                        mapobj, seenrev, user,
+                        notify=mapobj.notify and not notified,
+                    )
+                    notified = True
+
+        # Handle tags
+        for tag, (revision, branches) in tags.iteritems():
+            if not branches:
+                print("No branch found for tag '%s'" % tag)
+                continue
+            mapobjs = WebHookMapping.objects.filter(
+                repourl=self.url, branch__in=branches,
+            )
+            for mapobj in mapobjs:
+                seenrev, created = LastSeenRevision.objects.get_or_create(
+                    mapping=mapobj
+                )
+                if created or seenrev.revision != revision:
```

Comment:
What I meant is that the annotated tag object itself has a hash, but it references a commit with _another_ hash.  Iirc this always hits the `seenrev.revision != revision` condition.  Just a remark to be aware of that, we/I should add a unit test for this case.